File: ocaml/src/proxy_protocol.ml
(*
Copyright (C) 2016 iNuron NV

This file is part of Open vStorage Open Source Edition (OSE), as available from


    http://www.openvstorage.org and
    http://www.openvstorage.com.

This file is free software; you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
as published by the Free Software Foundation, in version 3 as it comes
in the <LICENSE.txt> file of the Open vStorage OSE distribution.

Open vStorage is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY of any kind.
*)

(* TODO:
   - remove std::exception from llio.cc?
 *)

open Prelude
open Stat
open Range_query_args2

module ProxyStatistics = struct
    include Stat


    module H = struct
        type ('a,'b) t = ('a *'b) list [@@deriving show, yojson]

        let h_to a_to b_to =
          Llio2.WriteBuffer.list_to
            (Llio2.WriteBuffer.pair_to
               a_to b_to)

        let h_from a_from b_from =
          Llio2.ReadBuffer.list_from
            (Llio2.ReadBuffer.pair_from
               a_from b_from)

        let find t a = List.assoc a t

        let add t a b = (a,b) :: t
      end

    type ns_t = {
        mutable upload: stat;
        mutable download: stat;
        mutable manifest_cached: int;
        mutable manifest_from_nsm  : int;
        mutable manifest_stale : int;
        mutable fragment_cache_hits: int;
        mutable fragment_cache_misses:int;

        mutable partial_read_size: stat;
        mutable partial_read_count: stat;
        mutable partial_read_time : stat;
        mutable partial_read_objects: stat;
      }[@@ deriving show, yojson]

    let ns_make () =
      { upload = Stat.make();
        download = Stat.make();
        manifest_cached = 0;
        manifest_from_nsm  = 0;
        manifest_stale = 0;
        fragment_cache_hits = 0;
        fragment_cache_misses = 0;

        partial_read_size    = Stat.make ();
        partial_read_count   = Stat.make ();
        partial_read_time    = Stat.make ();
        partial_read_objects = Stat.make ();
      }

    let ns_to buf t =
      let module Llio = Llio2.WriteBuffer in
      Stat_deser.to_buffer' buf t.upload;
      Stat_deser.to_buffer' buf t.download;
      Llio.int_to buf t.manifest_cached;
      Llio.int_to buf t.manifest_from_nsm;
      Llio.int_to buf t.manifest_stale;
      Llio.int_to buf t.fragment_cache_hits;
      Llio.int_to buf t.fragment_cache_misses;

      Stat_deser.to_buffer' buf t.partial_read_size;
      Stat_deser.to_buffer' buf t.partial_read_count;
      Stat_deser.to_buffer' buf t.partial_read_time;
      Stat_deser.to_buffer' buf t.partial_read_objects

    let ns_from buf =
      let module Llio = Llio2.ReadBuffer in
      let upload   = Stat_deser.from_buffer' buf in
      let download = Stat_deser.from_buffer' buf in
      let manifest_cached    = Llio.int_from buf in
      let manifest_from_nsm  = Llio.int_from buf in
      let manifest_stale     = Llio.int_from buf in
      let fragment_cache_hits    = Llio.int_from buf in
      let fragment_cache_misses  = Llio.int_from buf in
      (* trick to be able to work with <= 0.6.20 proxies *)
      let (partial_read_size,
           partial_read_count,
           partial_read_time,
           partial_read_objects)
        =
        begin
          if Llio.buffer_done buf
          then
            let r = Stat.make () in
            r,r,r,r
          else
            let s = Stat_deser.from_buffer' buf in
            let c = Stat_deser.from_buffer' buf in
            let t = Stat_deser.from_buffer' buf in
            if Llio.buffer_done buf
            then s,c,t,Stat.make()
            else
              let n = Stat_deser.from_buffer' buf in
              s,c,t,n
        end
      in

      { upload ; download;
        manifest_cached;
        manifest_from_nsm;
        manifest_stale;
        fragment_cache_hits;
        fragment_cache_misses;
        partial_read_size;
        partial_read_count;
        partial_read_time;
        partial_read_objects;
      }

    type t = {
        mutable creation:timestamp;
        mutable period: float;
        mutable ns_stats : (string, ns_t) H.t;
      } [@@deriving show, yojson]

    type t' = {
        t : t;
        changed_ns_stats : (string, unit) Hashtbl.t;
      }

    let make () =
      let creation = Unix.gettimeofday () in
      { t = { creation; period = 0.0;
              ns_stats = [];
            };
        changed_ns_stats = Hashtbl.create 3;
      }

    let to_buffer buf t =
      let module Llio = Llio2.WriteBuffer in
      let ser_version = 1 in Llio.int8_to buf ser_version;
      Llio.float_to buf t.creation;
      Llio.float_to buf t.period;
      H.h_to Llio.string_to ns_to buf t.ns_stats

    let from_buffer buf =
      let module Llio = Llio2.ReadBuffer in
      let ser_version = Llio.int8_from buf in
      assert (ser_version = 1);
      let creation = Llio.float_from buf in
      let period   = Llio.float_from buf in
      let ns_stats = H.h_from Llio.string_from ns_from buf in
      {creation;period;ns_stats}


    let deser = from_buffer, to_buffer

    let stop t = t.period <- Unix.gettimeofday() -. t.creation

    let clone t = { t.t with creation = t.t.creation }

    let clear t =
      Hashtbl.clear t.changed_ns_stats;
      t.t.creation <- Unix.gettimeofday ();
      t.t.period <- 0.0;
      t.t.ns_stats <- []

    let find t ns =
      Hashtbl.replace t.changed_ns_stats ns ();
      try H.find t.t.ns_stats ns
      with Not_found ->
        let v = ns_make () in
        let () = t.t.ns_stats <- H.add t.t.ns_stats ns v in
        v

    let show' ~only_changed t =
      show
        (if only_changed
         then
           { t.t with
             ns_stats =
               List.filter
                 (fun (namespace, _) -> Hashtbl.mem t.changed_ns_stats namespace)
                 t.t.ns_stats
           }
         else t.t)

    let clear_ns_stats_changed t =
      let r = Hashtbl.length t.changed_ns_stats in
      Hashtbl.clear t.changed_ns_stats;
      r

   let new_upload t ns delta =
     let ns_stats = find t ns in
     ns_stats.upload <- _update ns_stats.upload delta

   let incr_manifest_src ns_stats =
     let open Cache in
     function
     | Fast ->
          ns_stats.manifest_cached <- ns_stats.manifest_cached + 1
     | Slow ->
        ns_stats.manifest_from_nsm <- ns_stats.manifest_from_nsm + 1
     | Stale ->
        ns_stats.manifest_stale <- ns_stats.manifest_stale + 1

   let new_download t ns delta manifest_src (fg_hits, fg_misses) =
     let ns_stats = find t ns in
     let () =

       incr_manifest_src ns_stats manifest_src
     in
     let () = ns_stats.fragment_cache_hits <-
                ns_stats.fragment_cache_hits  + fg_hits
     in
     let () = ns_stats.fragment_cache_misses <-
                ns_stats.fragment_cache_misses + fg_misses
     in
     ns_stats.download <- _update ns_stats.download delta

   let new_read_objects_slices
         t ns
         ~total_length ~n_slices ~n_objects ~mf_sources
         ~fc_hits ~fc_misses
         ~took
     =
     let ns_stats = find t ns in
     ns_stats.partial_read_size    <- _update ns_stats.partial_read_size  (float total_length);
     ns_stats.partial_read_count   <- _update ns_stats.partial_read_count (float n_slices);
     ns_stats.partial_read_objects <- _update ns_stats.partial_read_objects (float n_objects);
     ns_stats.partial_read_time    <- _update ns_stats.partial_read_time  took;
     List.iter (incr_manifest_src ns_stats) mf_sources;
     ns_stats.fragment_cache_hits   <- ns_stats.fragment_cache_hits + fc_hits;
     ns_stats.fragment_cache_misses <- ns_stats.fragment_cache_misses + fc_misses;
     ()

end

module Protocol = struct

  let magic = 1148837403l
  let version = 1l

  module Amgrp = Albamgr_protocol.Protocol
  module Nsmhp = Nsm_host_protocol.Protocol
  module Nsmp = Nsm_protocol.Protocol

  module Namespace = Amgrp.Namespace

  type object_name = string[@@deriving show]

  type file_name = string

  type encryption_key = string option
  type overwrite = bool
  type may_not_exist = bool

  (* a call with return value has_more may return less than the
     requested amount of objects/namespaces. when all values in
     the range have been delivered has_more=false, otherwise
     has_more=true. *)
  type has_more = bool

  type preset_name = string
  type offset = Int64.t [@@deriving show]
  type length = int [@@deriving show]
  type data = string

  type consistent_read = bool
  type should_cache = bool

  type ('i, 'o) request =
    | ListNamespaces : (string RangeQueryArgs.t,
                        Namespace.name Std.counted_list * has_more) request
    | NamespaceExists : (Namespace.name, bool) request
    | CreateNamespace : (Namespace.name * preset_name option, unit) request
    | DeleteNamespace : (Namespace.name, unit) request

    | ListObjects : (Namespace.name *
                     string RangeQueryArgs.t,
                     object_name Std.counted_list * has_more) request
    | ReadObjectFs : (Namespace.name *
                      object_name *
                      file_name *
                      consistent_read *
                      should_cache,
                      unit) request
    | WriteObjectFs : (Namespace.name *
                       object_name *
                       file_name *
                       overwrite *
                       Checksum.Checksum.t option,
                       unit) request
    | DeleteObject : (Namespace.name *
                      object_name *
                      may_not_exist,
                      unit) request
    | GetObjectInfo : (Namespace.name *
                       object_name *
                       consistent_read *
                       should_cache,
                       Int64.t * Nsm_model.Checksum.t) request
    | ReadObjectsSlices : (Namespace.name *
                             (object_name * (offset * length) list) list *
                           consistent_read,
                           data) request
    | InvalidateCache : (Namespace.name, unit) request
    | DropCache : (Namespace.name, unit) request
    | ProxyStatistics : (bool, ProxyStatistics.t) request
    | GetVersion : (unit, (int * int * int * string)) request
    | OsdView : (unit, (string * Albamgr_protocol.Protocol.Osd.ClaimInfo.t) Std.counted_list
                       * (Albamgr_protocol.Protocol.Osd.id
                          * Nsm_model.OsdInfo.t
                          * Osd_state.t) Std.counted_list) request
    | GetClientConfig : (unit, Alba_arakoon.Config.t) request
    | Ping : (float, float) request

  type request' = Wrap : _ request -> request'
  let command_map = [ 1, Wrap ListNamespaces, "ListNamespaces";
                      2, Wrap NamespaceExists, "NamespaceExists";
                      3, Wrap CreateNamespace, "CreateNamespace";
                      4, Wrap DeleteNamespace, "DeleteNamespace";

                      5, Wrap ListObjects, "ListObjects";
                      8, Wrap DeleteObject, "DeleteObject";
                      9, Wrap GetObjectInfo, "GetObjectInfo";
                      10, Wrap ReadObjectFs, "ReadObjectFs";
                      11, Wrap WriteObjectFs, "WriteObjectFs";
                      13, Wrap ReadObjectsSlices, "ReadObjectsSlices";
                      14, Wrap InvalidateCache, "InvalidateCache";
                      15, Wrap ProxyStatistics, "ProxyStatistics";
                      16, Wrap DropCache, "DropCache";
                      17, Wrap GetVersion, "GetVersion";
                      18, Wrap OsdView,    "OsdView";
                      19, Wrap GetClientConfig, "GetClientConfig";
                      20, Wrap Ping, "Ping";
                    ]

  module Error = struct
    type t =
      | Unknown                 [@value 1]
      | OverwriteNotAllowed     [@value 2]
      | ObjectDoesNotExist      [@value 3]
      | NamespaceAlreadyExists  [@value 4]
      | NamespaceDoesNotExist   [@value 5]
      (* | EncryptionKeyRequired   [@value 6] *)
      | ChecksumMismatch        [@value 7]
      | ChecksumAlgoNotAllowed  [@value 8]
      | PresetDoesNotExist      [@value 9]
      | BadSliceLength          [@value 10]
      | OverlappingSlices       [@value 11]
      | SliceOutsideObject      [@value 12]
      | UnknownOperation        [@value 13]
      | FileNotFound            [@value 14]
      | NoSatisfiablePolicy     [@value 15]
      | ProtocolVersionMismatch [@value 17]
    [@@deriving show, enum]

    exception Exn of t

    let failwith err = raise (Exn err)

    let err2int = to_enum
    let int2err x = Option.get_some_default Unknown (of_enum x)
  end

  let wrap_unknown_operation f =
    try f ()
    with Not_found -> Error.(failwith UnknownOperation)

  let command_to_code =
    let hasht = Hashtbl.create 3 in
    List.iter (fun (code, comm, txt) -> Hashtbl.add hasht comm code) command_map;
    (fun comm -> wrap_unknown_operation (fun () -> Hashtbl.find hasht comm))

  let code_to_command =
    let hasht = Hashtbl.create 3 in
    List.iter (fun (code, comm, txt) -> Hashtbl.add hasht code comm) command_map;
    (fun code -> wrap_unknown_operation (fun () -> Hashtbl.find hasht code))

  let code_to_txt =
    let hasht = Hashtbl.create 3 in
    List.iter (fun (code, _, txt) -> Hashtbl.add hasht code txt) command_map;
    (fun code ->
     try Hashtbl.find hasht code with
     | Not_found -> Printf.sprintf "unknown operation %i" code)

  open Llio2
  let deser_request_i : type i o. (i, o) request -> i Deser.t = function
    | ListNamespaces -> RangeQueryArgs.deser' `MaxThenReverse Deser.string
    | NamespaceExists -> Deser.string
    | CreateNamespace -> Deser.tuple2 Deser.string (Deser.option Deser.string)
    | DeleteNamespace -> Deser.string

    | ListObjects ->
      Deser.tuple2
        Deser.string
        (RangeQueryArgs.deser' `MaxThenReverse Deser.string)
    | ReadObjectFs ->
      Deser.tuple5
        Deser.string
        Deser.string
        Deser.string
        Deser.bool
        Deser.bool
    | WriteObjectFs ->
      Deser.tuple5
        Deser.string
        Deser.string
        Deser.string
        Deser.bool
        (Deser.option Checksum_deser.deser')
    | DeleteObject ->
      Deser.tuple3
        Deser.string
        Deser.string
        Deser.bool
    | GetObjectInfo ->
      Deser.tuple4
        Deser.string
        Deser.string
        Deser.bool
        Deser.bool
    | ReadObjectsSlices ->
      Deser.tuple3
        Deser.string
        (Deser.list
           (Deser.pair
              Deser.string
              (Deser.list
                 (Deser.tuple2
                    Deser.int64
                    Deser.int))))
        Deser.bool
    | InvalidateCache -> Deser.string
    | DropCache -> Deser.string
    | ProxyStatistics -> Deser.bool
    | GetVersion      -> Deser.unit
    | OsdView         -> Deser.unit
    | GetClientConfig -> Deser.unit
    | Ping            -> Deser.float

  let deser_request_o : type i o. (i, o) request -> o Deser.t = function
    | ListNamespaces -> Deser.tuple2 (Deser.counted_list Deser.string) Deser.bool
    | NamespaceExists -> Deser.bool
    | CreateNamespace -> Deser.unit
    | DeleteNamespace -> Deser.unit

    | ListObjects -> Deser.tuple2 (Deser.counted_list Deser.string) Deser.bool
    | ReadObjectFs -> Deser.unit
    | WriteObjectFs -> Deser.unit
    | DeleteObject -> Deser.unit
    | GetObjectInfo -> Deser.tuple2 Deser.int64 Checksum_deser.deser'
    | ReadObjectsSlices -> Deser.string
    | InvalidateCache -> Deser.unit
    | DropCache -> Deser.unit
    | ProxyStatistics -> ProxyStatistics.deser
    | GetVersion -> Deser.tuple4
                      Deser.int
                      Deser.int
                      Deser.int
                      Deser.string
    | OsdView ->
       let deser_claim =
         Deser.counted_list
           (Deser.tuple2 Deser.string Osd_deser.ClaimInfo.deser) in
       let deser_osd   = Osd_deser.OsdInfo.from_buffer,
                         Osd_deser.OsdInfo.to_buffer
       in
       Deser.tuple2
         deser_claim
         (Deser.counted_list
            (Deser.tuple3 Deser.int32 deser_osd Osd_state.deser_state))
    | GetClientConfig ->
       Alba_arakoon_deser.Config.from_buffer, Alba_arakoon_deser.Config.to_buffer
    | Ping -> Deser.float
end

File: cpp/src/lib/proxy_protocol.cc
/*
Copyright (C) 2016 iNuron NV

This file is part of Open vStorage Open Source Edition (OSE), as available from


    http://www.openvstorage.org and
    http://www.openvstorage.com.

This file is free software; you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
as published by the Free Software Foundation, in version 3 as it comes
in the <LICENSE.txt> file of the Open vStorage OSE distribution.

Open vStorage is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY of any kind.
*/

#include "proxy_protocol.h"
#include <fstream>
#include <string.h>

#define _LIST_NAMESPACES 1
#define _NAMESPACE_EXISTS 2
#define _CREATE_NAMESPACE 3
#define _DELETE_NAMESPACE 4
#define _LIST_OBJECTS 5
#define _DELETE_OBJECT 8
#define _GET_OBJECT_INFO 9
#define _READ_OBJECT_FS 10
#define _WRITE_OBJECT_FS 11
//#define _READ_OBJECTS_SLICES 12 // deprecated.
#define _READ_OBJECTS_SLICES 13
#define _INVALIDATE_CACHE 14
#define _DROP_CACHE 16
#define _GET_PROXY_VERSION 17

#define _PING 20
#define _WRITE_OBJECT_FS2 21
#define _OSD_INFO 22
#define _READ_OBJECTS_SLICES2 23
#define _APPLY_SEQUENCE 24
#define _OSD_INFO2 28

namespace alba {
namespace proxy_protocol {

using llio::to;
using llio::from;

void write_tag(message_builder &mb, uint32_t tag) { to<uint32_t>(mb, tag); }

void read_status(message &m, Status &status) {
  uint32_t rc;
  from(m, rc);

  status.set_rc(rc);
  if (rc != 0) {
    string what;
    from(m, what);
    status._what = what;
  }
}

void write_range_params(message_builder &mb, const string &first,
                        const bool finc, const optional<string> &last,
                        const bool linc, const int max, const bool reverse) {
  to(mb, first);
  to(mb, finc);

  boost::optional<std::pair<string, bool>> lasto;
  if (boost::none == last) {
    lasto = boost::none;
  } else {
    lasto = std::pair<string, bool>(*last, linc);
  }
  to(mb, lasto);

  to<uint32_t>(mb, max);
  to(mb, reverse);
}

void write_list_namespaces_request(message_builder &mb, const string &first,
                                   const bool finc,
                                   const optional<string> &last,
                                   const bool linc, const int max,
                                   const bool reverse) {
  write_tag(mb, _LIST_NAMESPACES);
  write_range_params(mb, first, finc, last, linc, max, reverse);
}
void read_list_namespaces_response(message &m, Status &status,
                                   std::vector<string> &namespaces,
                                   bool &has_more) {
  read_status(m, status);
  if (status.is_ok()) {
    from(m, namespaces);
    from(m, has_more);
  }
}

void write_namespace_exists_request(message_builder &mb, const string &name) {
  write_tag(mb, _NAMESPACE_EXISTS);
  to(mb, name);
}
void read_namespace_exists_response(message &m, Status &status, bool &exists) {
  read_status(m, status);
  if (status.is_ok()) {
    from(m, exists);
  }
}

void write_create_namespace_request(message_builder &mb, const string &name,
                                    const optional<string> &preset_name) {
  write_tag(mb, _CREATE_NAMESPACE);
  to(mb, name);
  to(mb, preset_name);
}
void read_create_namespace_response(message &m, Status &status) {
  read_status(m, status);
}

void write_delete_namespace_request(message_builder &mb, const string &name) {
  write_tag(mb, _DELETE_NAMESPACE);
  to(mb, name);
}
void read_delete_namespace_response(message &m, Status &status) {
  read_status(m, status);
}

void write_list_objects_request(message_builder &mb, const string &namespace_,
                                const string &first, const bool finc,
                                const optional<string> &last, const bool linc,
                                const int max, const bool reverse) {
  write_tag(mb, _LIST_OBJECTS);
  to(mb, namespace_);
  write_range_params(mb, first, finc, last, linc, max, reverse);
}

void read_list_objects_response(message &m, Status &status,
                                std::vector<string> &objects, bool &has_more) {
  read_status(m, status);
  if (status.is_ok()) {
    from(m, objects);
    from(m, has_more);
  }
}

void write_read_object_fs_request(message_builder &mb, const string &namespace_,
                                  const string &object_name,
                                  const string &dest_file,
                                  const bool consistent_read,
                                  const bool should_cache) {
  write_tag(mb, _READ_OBJECT_FS);
  to(mb, namespace_);
  to(mb, object_name);
  to(mb, dest_file);
  to(mb, consistent_read);
  to(mb, should_cache);
}

void read_read_object_fs_response(message &m, Status &status) {
  read_status(m, status);
}

void _write_write_object_fs_request(message_builder &mb, uint32_t tag,
                                    const string &namespace_,
                                    const string &object_name,
                                    const string &input_file,
                                    const bool allow_overwrite,
                                    const Checksum *checksum) {
  write_tag(mb, tag);
  to(mb, namespace_);
  to(mb, object_name);
  to(mb, input_file);
  to(mb, allow_overwrite);
  if (nullptr == checksum) {
    to<boost::optional<const Checksum *>>(mb, boost::none);
  } else {
    to(mb, boost::optional<const Checksum *>(checksum));
  }
}

void write_write_object_fs_request(message_builder &mb,
                                   const string &namespace_,
                                   const string &object_name,
                                   const string &input_file,
                                   const bool allow_overwrite,
                                   const Checksum *checksum) {

  _write_write_object_fs_request(mb, _WRITE_OBJECT_FS, namespace_, object_name,
                                 input_file, allow_overwrite, checksum);
}

void write_write_object_fs2_request(message_builder &mb,
                                    const string &namespace_,
                                    const string &object_name,
                                    const string &input_file,
                                    const bool allow_overwrite,
                                    const Checksum *checksum) {

  _write_write_object_fs_request(mb, _WRITE_OBJECT_FS2, namespace_, object_name,
                                 input_file, allow_overwrite, checksum);
}

void read_write_object_fs_response(message &m, Status &status) {
  read_status(m, status);
}

void read_write_object_fs2_response(message &m, Status &status,
                                    ManifestWithNamespaceId &mf) {
  read_status(m, status);
  if (status.is_ok()) {

    from(m, mf);
  }
}

void write_delete_object_request(message_builder &mb, const string &namespace_,
                                 const string &object_name,
                                 const bool may_not_exist) {
  write_tag(mb, _DELETE_OBJECT);
  to(mb, namespace_);
  to(mb, object_name);
  to(mb, may_not_exist);
}

void read_delete_object_response(message &m, Status &status) {
  read_status(m, status);
}

void write_get_object_info_request(message_builder &mb,
                                   const string &namespace_,
                                   const string &object_name,
                                   const bool consistent_read,
                                   const bool should_cache) {
  write_tag(mb, _GET_OBJECT_INFO);
  to(mb, namespace_);
  to(mb, object_name);
  to(mb, consistent_read);
  to(mb, should_cache);
}

void read_get_object_info_response(message &m, Status &status, uint64_t &size,
                                   Checksum *&checksum) {
  read_status(m, status);
  if (status.is_ok()) {
    from(m, size);
    from(m, checksum);
  }
}

void write_read_objects_slices_request(message_builder &mb,
                                       const string &namespace_,
                                       const std::vector<ObjectSlices> &slices,
                                       const bool consistent_read) {
  write_tag(mb, _READ_OBJECTS_SLICES);
  to(mb, namespace_);
  to(mb, slices);
  to(mb, consistent_read);
}
void write_read_objects_slices2_request(message_builder &mb,
                                        const string &namespace_,
                                        const std::vector<ObjectSlices> &slices,
                                        const bool consistent_read) {
  write_tag(mb, _READ_OBJECTS_SLICES2);
  to(mb, namespace_);
  to(mb, slices);
  to(mb, consistent_read);
}

void read_read_objects_slices_response(
    message &m, Status &status,
    const std::vector<ObjectSlices> &objects_slices) {
  read_status(m, status);
  if (status.is_ok()) {
    uint32_t size;
    from<uint32_t>(m, size);

    for (auto &object_slices : objects_slices) {
      for (auto &slice : object_slices.slices) {
        memcpy(slice.buf, m.current(slice.size), slice.size);
        m.skip(slice.size);
      }
    }
  }
}

void _read_object_infos(message &m, std::vector<object_info> &object_infos) {
  // todo: automatically via templates
  uint32_t size;
  from(m, size);
  object_infos.resize(size);
  for (int32_t i = size - 1; i >= 0; --i) {
    std::string name;
    from(m, name);
    std::string future;
    from(m, future);
    std::unique_ptr<ManifestWithNamespaceId> umf(new ManifestWithNamespaceId());
    from(m, *umf);
    assert(name == umf->name);
    auto t =
        std::make_tuple(std::move(name), std::move(future), std::move(umf));
    object_infos[i] = std::move(t);
  }
}

void read_read_objects_slices2_response(
    message &m, Status &status, const std::vector<ObjectSlices> &objects_slices,
    std::vector<object_info> &object_infos) {
  read_status(m, status);
  if (status.is_ok()) {
    uint32_t size;
    from<uint32_t>(m, size);

    for (auto &object_slices : objects_slices) {
      for (auto &slice : object_slices.slices) {
        memcpy(slice.buf, m.current(slice.size), slice.size);
        m.skip(slice.size);
      }
    }

    _read_object_infos(m, object_infos);
  }
}

void write_apply_sequence_request(
    message_builder &mb, const string &namespace_, const bool write_barrier,
    const std::vector<std::shared_ptr<alba::proxy_client::sequences::Assert>>
        &asserts,
    const std::vector<std::shared_ptr<alba::proxy_client::sequences::Update>>
        &updates) {
  write_tag(mb, _APPLY_SEQUENCE);
  to(mb, namespace_);
  to(mb, write_barrier);
  to(mb, asserts);
  to(mb, updates);
}

void read_apply_sequence_response(message &m, Status &status,
                                  std::vector<object_info> &object_infos) {
  read_status(m, status);

  if (status.is_ok()) {
    _read_object_infos(m, object_infos);
  }
}

void write_invalidate_cache_request(message_builder &mb,
                                    const string &namespace_) {
  write_tag(mb, _INVALIDATE_CACHE);
  to(mb, namespace_);
}

void read_invalidate_cache_response(message &m, Status &status) {
  read_status(m, status);
}

void write_drop_cache_request(message_builder &mb, const string &namespace_) {
  write_tag(mb, _DROP_CACHE);
  to(mb, namespace_);
}

void read_drop_cache_response(message &m, Status &status) {
  read_status(m, status);
}

void write_get_proxy_version_request(message_builder &mb) {
  write_tag(mb, _GET_PROXY_VERSION);
}

void read_get_proxy_version_response(message &m, Status &status, int32_t &major,
                                     int32_t &minor, int32_t &patch,
                                     std::string &hash) {
  read_status(m, status);
  if (status.is_ok()) {
    from(m, major);
    from(m, minor);
    from(m, patch);
    from(m, hash);
  }
}

void write_ping_request(message_builder &mb, const double delay) {
  write_tag(mb, _PING);
  to(mb, delay);
}

void read_ping_response(message &m, Status &status, double &timestamp) {
  read_status(m, status);
  if (status.is_ok()) {
    from(m, timestamp);
  }
}

void write_osd_info_request(message_builder &mb) { write_tag(mb, _OSD_INFO); }

void _read_osd_infos(message &m, osd_map_t &result) {
  uint32_t n;
  from(m, n);
  for (uint32_t i = 0; i < n; i++) {
    osd_t osd_id;
    from(m, osd_id);
    std::string info_s;
    from(m, info_s);
    std::vector<char> mv(info_s.begin(), info_s.end());
    llio::message m2(mv);
    OsdInfo info;
    from(m2, info);
    OsdCapabilities caps;
    from(m, caps);
    auto p = std::shared_ptr<info_caps>(
        new info_caps(std::move(info), std::move(caps)));

    result[osd_id] = std::move(p);
  }
}
void read_osd_info_response(message &m, Status &status, osd_map_t &result) {
  read_status(m, status);
  if (status.is_ok()) {
    _read_osd_infos(m, result);
  }
}

void write_osd_info2_request(message_builder &mb) {
  ALBA_LOG(DEBUG, "write_osd_info2");
  write_tag(mb, _OSD_INFO2);
}

void read_osd_info2_response(message &m, Status &status,
                             rora_osd_map_t &result) {
  read_status(m, status);
  if (status.is_ok()) {
    uint32_t n;
    from(m, n);
    ALBA_LOG(DEBUG, "n=" << n);
    for (uint32_t i = 0; i < n; i++) {
      alba_id_t alba_id;
      from(m, alba_id);
      ALBA_LOG(DEBUG, "alba_id = " << alba_id);
      osd_map_t infos;
      _read_osd_infos(m, infos);
      auto entry = std::make_pair(std::move(alba_id), std::move(infos));
      result.push_back(std::move(entry));
    }
  }
}

std::ostream &operator<<(std::ostream &os, const SliceDescriptor &s) {
  os << "{ offset = " << s.offset << " , size = " << s.size << " }";
  return os;
}
std::ostream &operator<<(std::ostream &os, const ObjectSlices &s) {
  os << "{ object_name = ";
  dump_string(os, s.object_name);
  os << "slices = [ ";
  for (auto &sd : s.slices) {
    os << sd << ";";
  }
  os << " ] }";
  return os;
}

} // namespace

namespace llio {
template <>
void to(message_builder &mb,
        const proxy_protocol::SliceDescriptor &desc) noexcept {
  to(mb, desc.offset);
  to(mb, desc.size);
}
template <>
void to(message_builder &mb,
        const proxy_protocol::ObjectSlices &slices) noexcept {
  to(mb, slices.object_name);
  to(mb, slices.slices);
}
}
}
